_websiteutils/make_redirects_links.py

```diff
@@ -0,0 +1,202 @@
+                os.remove('./'+fullname)
+                if name.endswith(('.htm', '.html')):
+                    # make an html redirect.
+                    _log.info(f'Rewriting HTML: {fullname} in {last}')
+                    with open(fullname, 'w') as fout:
+                        oldname = '/' + os.path.join(last, fullname)
+                        st = html_redirect % (oldname, oldname, oldname)
+                        fout.write(st)
+                else:
+                    # soft link
+                    # Need to do these relative to where the link is
+                    # so if it is a level down `ln -s ../3.1.1/boo/who boo/who`
+                    last = os.path.join('..', last)
+                    depth = root.count('/')
+                    for i in range(depth):
+                        last = os.path.join('..', last)
+                    oldname = os.path.join(last, fullname)
+                    _log.info(f'Linking {fullname} to {oldname}')
+                    os.symlink(oldname, fullname)
+        for d in dirs:
+            do_links(d)
+
+
+def do_canonicals(dname):
+    """
+    For each html file in the versioned docs, make the canonical link point
+    to the newest version.
+    """
+    _log.debug(f'Walking {dname}')
+    for root, dirs, files in os.walk(dname):
+        for name in files:
+            fullname = os.path.join(root, name)
+            p = pathlib.Path(fullname)
+            _log.debug(f'Checking {fullname}')
+            if name.endswith(('.htm', '.html')):
+                basename = pathlib.Path(*p.parts[1:])
+                last = findlast(basename, tocheck)
+                if last is not None:
+                    update_canonical(fullname, last)
+
+        for d in dirs:
+            _log.info(f'DIR: {d}')
+            do_canonicals(os.path.join(dname,d))
+
+
+def update_canonical(fullname, last):
+    """
+    Change the canonical link in *fullname* to the same link in the
+    version given by *last*.  We do this with a regexp to prevent
+    removing any other content on a line that has the canonical link.
+
+    Note that if for some reason there are more than one canonical link
+    this will change  all of them.
+    """
+    p = pathlib.Path(fullname)
+    pre = 'https://matplotlib.org/'
+    pnew = pathlib.Path(last, *p.parts[1:])
+    newcanon = f'{pre+str(pnew)}'
+    _log.info(f'{p} to {pre+str(pnew)}')
+    with tempfile.NamedTemporaryFile(delete=False) as fout:
+        with open(fullname,'r') as fin:
+            for line in fin:
+                if '<link rel="canonical"' in line:
+                    ll = re.sub('<link rel="canonical" href=".*"',
+                                f'<link rel="canonical" href="{newcanon}"',
+                                line)
+                    _log.debug(f'new {line}->{ll}')
+                    fout.write(bytes(ll, encoding='utf-8'))
+                else:
+                    fout.write(bytes(line, encoding = 'utf-8'))
```

Comment:
Why not open `fin` in binary mode as well? Then only `newcanon` needs to be encoded.

Reply:
I did this because you asked, but I'm not sure its an improvement.  No doubt I am missing some tricks to dealing with byte strings....

Reply:
Ah, I forgot there's no `bytes` f-string equivalent; maybe encoding `new` instead would be better?

Reply:
Yep that works nicer thanks!